### PR TITLE
Bump stable PHP version used in CI

### PR DESCRIPTION
**Pull request by pierlon**
## Summary

Bumps the PHP version used in our CI test suite to 8.0.

I was also attempting to include PHP 8.1 into the WP integration test matrix but it's becoming a lot more complex than initially thought. It would require us to use the polyfilled test fixtures provided by the [PHPUnit-Polyfills](https://github.com/Yoast/PHPUnit-Polyfills/#testcases) library, but that would mean the WP test suite would also need to extend that library as well, which is the case for WP trunk. This has also been recently backported to WP 5.2 - 5.8 in https://github.com/WordPress/wordpress-develop/pull/1588. However, there is not much interest in further backporting those changes to older WP versions as yet (see https://github.com/WordPress/wordpress-develop/pull/1589), but I've filed a feature request in the WP Test Utils repo (https://github.com/Yoast/wp-test-utils/issues/14) to add a wrapper for those test fixtures. That should then allow us to use those test fixtures without worrying about the WP version being tested against.

Closes #6868 

## Checklist

- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [ ] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

File: tests/php/src/Admin/OptionsMenuTest.php [hunk 1]
```diff
@@ -198,6 +198,8 @@ public function get_can_validate_data() {
 	 * @param bool $can_validate
 	 */
 	public function test_enqueue_assets_right_hook_suffix( $can_validate ) {
+		$initial_action_count = did_action( 'amp_register_polyfills' );
```
Comment (westonruter) [inline tests/php/src/Admin/OptionsMenuTest.php]: The fact that this is not always initially `0` indicates the test suite is not cleaning itself up always.

File: tests/php/src/PairedRoutingTest.php [hunk 2]
```diff
@@ -901,6 +901,8 @@ function () {
 	 * @param callable|null $url_callback
 	 */
 	public function test_has_endpoint( $setup_callback, $expected_has_endpoint, $url_callback ) {
+		$GLOBALS['wp_the_query'] = $GLOBALS['wp_query']; // Needed as of <https://github.com/ampproject/amp-wp/pull/6466>.
```
Comment (westonruter) [inline tests/php/src/PairedRoutingTest.php]: Is the the base test case not doing this anymore?

File: tests/php/src/Validation/ValidationCountsRestControllerTest.php [hunk 3]
```diff
@@ -57,6 +57,7 @@ public function test_get_items_permissions_check() {
 		$this->assertWPError( $this->controller->get_items_permissions_check( new WP_REST_Request( 'GET', '/amp/v1/unreviewed-validation-counts' ) ) );
 
 		wp_set_current_user( $admin_user->ID );
+		update_user_meta( $admin_user->ID, UserAccess::USER_FIELD_DEVELOPER_TOOLS_ENABLED, wp_json_encode( false ) );
```
Comment (westonruter) [inline tests/php/src/Validation/ValidationCountsRestControllerTest.php]: I don't know why this is needed now.

File: tests/php/src/Validation/ScannableURLsRestControllerTest.php [hunk 4]
```diff
@@ -77,6 +79,9 @@ public function test_get_items_permissions_check() {
 	 * @covers ::get_item_schema()
 	 */
 	public function test_get_items() {
+		wp_set_current_user( self::factory()->user->create( [ 'role' => 'administrator' ] ) ); // Because the setting was registered and not cleaned up?
+		AMP_Options_Manager::update_option( Option::READER_THEME, ReaderThemes::DEFAULT_READER_THEME );
+		AMP_Options_Manager::update_option( Option::THEME_SUPPORT, AMP_Theme_Support::READER_MODE_SLUG );
```
Comment (westonruter) [inline tests/php/src/Validation/ScannableURLsRestControllerTest.php]: I'm not sure why the options need to be set now where they didn't before. Again, seems to indicate the database not being reset. Also, the need to set the current user indicates the setting was registered and thus the capability check is being done.

**Comment by westonruter** [comment]
@schlessera I'm getting a PHP error in `ServiceBasedPluginTest`:

```
3) AmpProject\AmpWP\Tests\Infrastructure\ServiceBasedPluginTest::test_it_can_be_registered
ParseError: syntax error, unexpected token "match"

/tmp/wordpress/src/wp-content/plugins/amp/tests/php/src/Infrastructure/ServiceBasedPluginTest.php:39
```

**Comment by schlessera** [comment]
@westonruter This is due to https://github.com/sebastianbergmann/phpunit/issues/4373, which is fixed with PHPUnit 9.3+, but [the test suite seems to use PHPUnit 9.2.6](https://github.com/ampproject/amp-wp/runs/5504424621?check_suite_focus=true#step:11:62) for the PHP 8.0 test run.

The reason why PHPUnit seems to be locked at 9.2 only is because of the version constraint of `phpdocumentor/reflection`. If that version is bumped to `^4` or `^5`, a newer version of PHPUnit is used for PHP 8, which fixes the issue.

**Comment by westonruter** [comment]
@schlessera Looks like you added that in 4721180f5bc682d73028cde6a4de89059afe7632 for the docs generation, and v3 of phpdocumentor/reflection was the last version that supported PHP 5.x. Both v4 and v5 both require PHP 7. I think I have a workaround for testing in cf82a91.

**Comment by westonruter** [comment]
Well, I'm about at the end of my rope trying to get PHPUnit to run on the various versions of WP and PHP. I think this needs someone else with a fresh pair of eyes.

**Comment by schlessera** [comment]
@westonruter PHPUnit is already being managed and properly required by the Yoast test package. It should have been enough to just get rid of the blocking dependency. Am I missing something?

**Comment by westonruter** [comment]
> PHPUnit is already being managed and properly required by the Yoast test package. It should have been enough to just get rid of the blocking dependency. Am I missing something?

@schlessera I didn't realize that. I recall I did try removing manual installation of PHPUnit but then the bootstrap failed with the `WP_CLI\Loggers\Regular` class not being found. Trying again in 84f2604.

File: tests/php/src/Admin/ReenableCssTransientCachingAjaxActionTest.php [hunk 5]
```diff
@@ -200,5 +199,8 @@ static function () {
 			$final_value,
 			AMP_Options_Manager::get_option( Option::DISABLE_CSS_TRANSIENT_CACHING )
 		);
+
+		remove_filter( 'wp_doing_ajax', '__return_true' );
+		remove_filter( 'wp_die_ajax_handler', $wp_die_ajax_handler );
```
Comment (westonruter) [inline tests/php/src/Admin/ReenableCssTransientCachingAjaxActionTest.php]: @dhaval-parekh Should this be needed since tearDown/setUp should reset all filters to their original state?

**Comment by westonruter** [comment]
Well, tests are passing now. The last issue is the Codecov integration:

image

File: tests/php/src/Admin/ReenableCssTransientCachingAjaxActionTest.php (hunk 5 above)
Comment (dhaval-parekh) [reply tests/php/src/Admin/ReenableCssTransientCachingAjaxActionTest.php]: Yes, We can remove it. ( It was added to check whether that is causing the issue or not. )

I have removed it in [be8e3bf](https://github.com/ampproject/amp-wp/pull/6466/commits/be8e3bfdb8516fd326b89bc31a4ab07c00e93db0)

File: tests/php/src/ReaderThemeLoaderTest.php [hunk 6]
```diff
@@ -56,6 +56,7 @@ public function test_is_enabled() {
 		$this->assertTrue( $this->instance->is_enabled() );
 		$this->assertNotEquals( get_template(), $reader_theme_slug );
 
+		$GLOBALS['wp_the_query'] = $GLOBALS['wp_query']; // Needed as of <https://github.com/ampproject/amp-wp/pull/6466>.
```
Comment (westonruter) [inline tests/php/src/ReaderThemeLoaderTest.php]: As above, not sure why I needed to add this.

File: tests/php/src/ReaderThemeLoaderTest.php [hunk 7]
```diff
@@ -214,6 +215,7 @@ public function test_override_theme() {
 		if ( ! wp_get_theme( $active_theme_slug )->exists() || ! wp_get_theme( $reader_theme_slug )->exists() ) {
 			$this->markTestSkipped();
 		}
+		wp_reset_query(); // phpcs:ignore WordPress.WP.DiscouragedFunctions.wp_reset_query_wp_reset_query
```
Comment (westonruter) [inline tests/php/src/ReaderThemeLoaderTest.php]: Why was this needed?

File: tests/php/src/ReaderThemeSupportFeaturesTest.php [hunk 8]
```diff
@@ -247,6 +248,7 @@ public function test_handle_theme_update_with_reader_theme_enabled_and_overridin
 		AMP_Options_Manager::update_option( Option::PRIMARY_THEME_SUPPORT, null );
 		switch_theme( self::THEME_PRIMARY );
 		AMP_Options_Manager::update_option( Option::READER_THEME, self::THEME_READER );
+		AMP_Options_Manager::update_option( Option::THEME_SUPPORT, AMP_Theme_Support::READER_MODE_SLUG );
```
Comment (westonruter) [inline tests/php/src/ReaderThemeSupportFeaturesTest.php]: Reader mode should have been the default. Curious why I needed to add this.

File: tests/php/validation/test-class-amp-validation-error-taxonomy.php [hunk 9]
```diff
@@ -31,14 +31,16 @@ class Test_AMP_Validation_Error_Taxonomy extends TestCase {
 	/**
 	 * Resets the state after each test method.
 	 */
-	public function tearDown() {
+	public function tear_down() {
 		$_REQUEST = [];
 		AMP_Options_Manager::update_option( Option::THEME_SUPPORT, AMP_Theme_Support::READER_MODE_SLUG );
+		remove_theme_support( 'amp' );
```
Comment (westonruter) [inline tests/php/validation/test-class-amp-validation-error-taxonomy.php]: Why was this necessary?

File: tests/php/validation/test-class-amp-validation-manager.php [hunk 10]
```diff
@@ -581,7 +581,6 @@ public function test_override_validation_error_statuses_with_good_nonce() {
 	 * Test validation error overrides for when bad nonce is supplied.
 	 *
 	 * @covers AMP_Validation_Manager::override_validation_error_statuses()
-	 * @expectedException WPDieException
```
Comment (westonruter) [inline tests/php/validation/test-class-amp-validation-manager.php:584]: @dhaval-parekh Are these annotations no longer supported?

File: tests/php/src/ReaderThemeLoaderTest.php (hunk 7 above)
Comment (westonruter) [reply tests/php/src/ReaderThemeLoaderTest.php]: I guess same reason that this is needed above: `$GLOBALS['wp_the_query'] = $GLOBALS['wp_query']`

File: tests/php/src/Validation/ValidationCountsRestControllerTest.php (hunk 3 above)
Comment (westonruter) [reply tests/php/src/Validation/ValidationCountsRestControllerTest.php]: When running the test in isolation, the options are:

```
Isolated:
array(15) {
  ["theme_support"]=>
  string(6) "reader"
  ["supported_post_types"]=>
  array(2) {
    [0]=>
    string(4) "post"
    [1]=>
    string(4) "page"
  }
  ["analytics"]=>
  array(0) {
  }
  ["all_templates_supported"]=>
  bool(true)
  ["supported_templates"]=>
  array(1) {
    [0]=>
    string(11) "is_singular"
  }
  ["version"]=>
  string(11) "2.3.0-alpha"
  ["reader_theme"]=>
  string(6) "legacy"
  ["paired_url_structure"]=>
  string(9) "query_var"
  ["plugin_configured"]=>
  bool(false)
  ["delete_data_at_uninstall"]=>
  bool(true)
  ["mobile_redirect"]=>
  bool(true)
  ["late_defined_slug"]=>
  NULL
  ["suppressed_plugins"]=>
  array(0) {
  }
  ["sandboxing_enabled"]=>
  bool(false)
  ["sandboxing_level"]=>
  int(1)
}
```

When running all the tests, the options then are at this point:

```
array(15) {
  ["theme_support"]=>
  string(12) "transitional"
  ["supported_post_types"]=>
  array(2) {
    [0]=>
    string(4) "post"
    [1]=>
    string(4) "page"
  }
  ["analytics"]=>
  array(0) {
  }
  ["all_templates_supported"]=>
  bool(true)
  ["supported_templates"]=>
  array(1) {
    [0]=>
    string(11) "is_singular"
  }
  ["version"]=>
  string(11) "2.3.0-alpha"
  ["reader_theme"]=>
  string(6) "legacy"
  ["paired_url_structure"]=>
  string(9) "query_var"
  ["plugin_configured"]=>
  bool(false)
  ["delete_data_at_uninstall"]=>
  bool(true)
  ["mobile_redirect"]=>
  bool(true)
  ["late_defined_slug"]=>
  NULL
  ["suppressed_plugins"]=>
  array(0) {
  }
  ["sandboxing_enabled"]=>
  bool(false)
  ["sandboxing_level"]=>
  int(1)
}
```

The diff being:

```diff
3c3
<   string(6) "reader"
---
>   string(12) "transitional"
```

So yeah, for some reason the options are not getting reset.

Comment (westonruter) [reply tests/php/src/Validation/ValidationCountsRestControllerTest.php]: When running the test is isolated, calling `AMP_Theme_Support::get_theme_support_args()` returns `false`. 

When not in isolation, the function returns:

```
 array(1) {
  ["paired"]=>
  bool(true)
}
```

So the issue seems to be that the `$_wp_theme_features` global is not getting reset.

File: tests/php/validation/test-class-amp-validation-error-taxonomy.php (hunk 9 above)
Comment (westonruter) [reply tests/php/validation/test-class-amp-validation-error-taxonomy.php]: Addressed in 774050fac2a2a57aac567f0f5dbce89a32cdd4aa

File: tests/php/src/ReaderThemeSupportFeaturesTest.php (hunk 8 above)
Comment (westonruter) [reply tests/php/src/ReaderThemeSupportFeaturesTest.php]: Addressed in 774050fac2a2a57aac567f0f5dbce89a32cdd4aa

**Review by westonruter — APPROVED** [review]
😅 

Thank you @pierlon and @dhaval-parekh for this!

**Comment by westonruter** [comment]
A hard fought battle:

image

File: tests/php/validation/test-class-amp-validation-manager.php (hunk 10 above)
Comment (dhaval-parekh) [reply tests/php/validation/test-class-amp-validation-manager.php:584]: This annotation is deprecated and will be removed in PHPUnit 9.
